### PR TITLE
texlab: init at 1.6.0

### LINKED ISSUE #67934: packaging request: texlab
**Project description**
An LSP server for latex. Rust + node I think

**Metadata**

* homepage URL: https://github.com/latex-lsp/texlab
* source URL: https://github.com/latex-lsp/texlab
* license: mit
* platforms: unix, linux, darwin, ...

### PR DESCRIPTION
**Pull request by kira-bruneau**
###### Motivation for this change
Fixes https://github.com/NixOS/nixpkgs/issues/67934

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
```
[118 built, 1782 copied (4690.2 MiB), 1542.7 MiB DL]
19 package were build:
antora create-cycle-app emojione hueadm image_optim iosevka joplin lumo parity-ui rambox-pro sage sageWithDoc slack slack-dark texlab thelounge triton twemoji-color-font wasm-text-gen
```
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Notify maintainers

cc @doronbehar

### CODE REVIEW

**Comment by MetaDark** [comment]
All unit tests pass and I manually verified the binary using Emacs as a client.

@doronbehar I added you as a maintainer, is that ok with you?

**Comment by doronbehar** [comment]
Looks good! I tried to figure this out last night with no luck. I have 1 comment:

What I was trying to do, based on other packages I saw that use node2nix, was writing a script such as this:

```sh
#!/usr/bin/env nix-shell
#! nix-shell -i bash -p nodePackages.node2nix jq

set -eu -o pipefail

if [ "$#" -ne 1 ] || [[ "$1" == -* ]]; then
	echo "Usage: $0 <git release tag>"
	exit 1
fi

TEXLAB_WEB_SRC="https://raw.githubusercontent.com/latex-lsp/texlab/$1"
# We use curl and `jq` inorder to make the `devDependencies` normal `dependencies`
curl --silent "$TEXLAB_WEB_SRC/src/citeproc/js/package.json" | \
	jq '. + {"dependencies": .devDependencies} | del(.devDependencies)' > package.json

wget "$TEXLAB_WEB_SRC/src/citeproc/js/package-lock.json" -O package-lock.json

node2nix \
	--nodejs-10
	-i package.json -l package-lock.json \
	--composition citeproc-packages.nix \
	-e ../../../development/node-packages/node-env.nix
```

However, from the diff of this PR it seems that you added a `package.json` from the repo and ran the node2nix script of all node packages. I'm talking about this script:

https://github.com/NixOS/nixpkgs/blob/5e70be026ee4eb05ed1c25d659848ab2f7102100/pkgs/development/node-packages/generate.sh

Right?

Maybe we can at least take the idea of this script to automate the `package.json` and `package-lock.json` download procedure?

Oh and also, the PR doesn't use `package-lock.json` now, right? I think we should..

> @doronbehar I added you as a maintainer, is that ok with you?

Yes!

**Comment by MetaDark** [comment]
@doronbehar

> However, from the diff of this PR it seems that you added a package.json from the repo and ran the node2nix script of all node packages. I'm talking about this script:
>
> https://github.com/NixOS/nixpkgs/blob/5e70be026ee4eb05ed1c25d659848ab2f7102100/pkgs/development/node-packages/generate.sh
>
> Right?
>
> Maybe we can at least take the idea of this script to automate the package.json and package-lock.json download procedure?

I followed the same style used in luma and iosveka for grabbing nodejs build dependencies. The `package.json` I added for citeproc is only a projection of the one stored in the texlab repo. It doesn't contain `prettier`, `tslint`, or `tslint-config-prettier` since they aren't required for running `npm run dist`. This will add some extra maintenance overhead for each update and probably shouldn't be automated with a script, but I think it's worth it to reduce the closure size.

> Oh and also, the PR doesn't use package-lock.json now, right? I think we should..

I didn't include a lock file for a few reasons:

- `generate.sh` isn't setup to use lock files. It would have to be refactored.
- From what I understand, package-lock.json is only used for creating a reproducible environment for development / CI. If citeproc were published to npm, the lock file would be ignored.
- The nix expressions for dependencies can be shared with other node packages that match the same semver.

**Comment by doronbehar** [comment]
> > Oh and also, the PR doesn't use package-lock.json now, right? I think we should..
> I didn't include a lock file for a few reasons:
> 
> * The files generated by node2nix still allow for a reproducible environment (however it may not be the exact same reproducible environment).
> * The nix expressions for dependencies can be shared with other node packages that match the same semver pattern.

I see :+1:.

> I followed the same style used in luma and iosveka for grabbing nodejs build dependencies. The package.json I added for citeproc is only a projection of the one stored in the texlab repo. It doesn't contain prettier, tslint, or tslint-config-prettier since they aren't required for running npm run dist. This will add some extra maintenance overhead for each update and probably shouldn't be automated with a script, but I think it's worth it to reduce the closure size.

That's a good reason for not including those linting dependencies but according to `git grep`, the node packages `prettier`, `tslint` etc. are already present in `node-packages-v10.nix` for example. I think it wouldn't be that significant if we'll download (only) the `package.json` as is from upstream with a script such as:


```patch
diff --git c/pkgs/development/tools/misc/texlab/citeproc/update-package.json.sh i/pkgs/development/tools/misc/texlab/citeproc/update-package.json.sh
new file mode 100755
index 00000000000..54fbfaccaff
--- /dev/null
+++ i/pkgs/development/tools/misc/texlab/citeproc/update-package.json.sh
@@ -0,0 +1,14 @@
+#!/usr/bin/env nix-shell
+#! nix-shell -i bash jq
+
+set -eu -o pipefail
+
+if [ "$#" -ne 1 ] || [[ "$1" == -* ]]; then
+	echo "Usage: $0 <git release tag>"
+	exit 1
+fi
+
+TEXLAB_WEB_SRC="https://raw.githubusercontent.com/latex-lsp/texlab/$1"
+
+curl --silent "$TEXLAB_WEB_SRC/src/citeproc/js/package.json" | \
+	jq '. + {"dependencies": .devDependencies} | del(.devDependencies)' > package.json
```

**Comment by MetaDark** [comment]
@doronbear That's a good point :+1:. I will try using your script.

**Review by doronbehar — APPROVED** [review]
Looks great now!

**Comment by veprbl** [comment]
Need to resolve the merge conflict. Preferably by not applying irrelevant changes to node-packages-vXX.nix.

**Comment by MetaDark** [comment]
@veprbl The `node-packages-vXX.nix` files are generated from a script. Should I just selectively exclude changes not used by texlab?

**Comment by veprbl** [comment]
@MetaDark Yes, that's my advice. This should help to limit the PR bitrot until it's reviewed/merged.

**Comment by doronbehar** [comment]
@veprbl Do you conceive a way to make sure the conflicts will always be resolved without making manual changes to the auto generated "Do not edit!" `node-packages-v##.nix` files? I tend to think there isn't a safe way to do it. Hence the only way this PR could be merged is if someone with merge rights will merge it as soon as it will be updated (by @MetaDark running once more `node2nix`)..

**Comment by MetaDark** [comment]
I ran into problems trying to manually removed unneeded changes, so I just decided to rebuild the whole expression again.

I started building a tool to automate this, but it's not complete yet.

**Comment by doronbehar** [comment]
I hope there won't be a need for that, perhaps @marsam would be willing to review and merge this quickly? As this is a bit urgent, it will be harder to merge this once there will be merge conflicts..

File: pkgs/development/tools/misc/texlab/default.nix
Comment (marsam) [inline pkgs/development/tools/misc/texlab/default.nix]: would you mind moving this to the `preBuild` below?

File: pkgs/top-level/all-packages.nix
Comment (marsam) [inline pkgs/top-level/all-packages.nix]: please add the following to add support for darwin:
```diff
diff --git i/pkgs/development/tools/misc/texlab/default.nix w/pkgs/development/tools/misc/texlab/default.nix
index c97131a4794..19ce67de803 100644
--- i/pkgs/development/tools/misc/texlab/default.nix
+++ w/pkgs/development/tools/misc/texlab/default.nix
@@ -1,7 +1,8 @@
-{ lib
+{ stdenv
 , rustPlatform
 , fetchFromGitHub
 , nodejs
+, Security
 , texlab-citeproc-build-deps
 }:
 
@@ -20,6 +21,8 @@ rustPlatform.buildRustPackage rec {
 
   nativeBuildInputs = [ nodejs ];
 
+  buildInputs = stdenv.lib.optionals stdenv.isDarwin [ Security ];
+
   patchPhase = ''
     rm build.rs
     ln -s ${texlab-citeproc-build-deps}/lib/node_modules/citeproc/node_modules src/citeproc/js
@@ -27,7 +30,7 @@ rustPlatform.buildRustPackage rec {
 
   preBuild = "(cd src/citeproc/js && npm run dist)";
 
-  meta = with lib; {
+  meta = with stdenv.lib; {
     description = "An implementation of the Language Server Protocol for LaTeX";
     homepage = https://texlab.netlify.com/;
     license = licenses.mit;
diff --git i/pkgs/top-level/all-packages.nix w/pkgs/top-level/all-packages.nix
index ecaae36923a..a5fe4201441 100644
--- i/pkgs/top-level/all-packages.nix
+++ w/pkgs/top-level/all-packages.nix
@@ -10383,6 +10383,7 @@ in
   texi2mdoc = callPackage ../tools/misc/texi2mdoc { };
 
   texlab = callPackage ../development/tools/misc/texlab {
+    inherit (darwin.apple_sdk.frameworks) Security;
     texlab-citeproc-build-deps = nodePackages."texlab-citeproc-build-deps-../tools/misc/texlab/citeproc";
   };
 ```

**Review by marsam — APPROVED** [review]

**Comment by marsam** [comment]
Thanks for working in this, and sorry for the delay
>I also started working on a tool to try to support partial builds. It's not complete yet, but it could be helpful in the future.

I would also like to work on this in the future, right now `node-packages/node-packages-vXX.nix` only lists the packages without version, I was thinking that we could keep a package.json and package-lock.json in nixpkgs to track only the related updates

**Comment by MetaDark** [comment]
@marsam That definitely sounds simpler than what I was originally planning. I was thinking of parsing `node-packages.nix`. Although, the array style used in `node-packages-vXX.json` isn't part of the package.json specification: https://docs.npmjs.com/files/package.json. So it would probably have to be generated manually instead of through npm.

**Comment by veprbl** [comment]
Perhaps this should have gone into a separate node2nix package set like it's done in other instances:

- https://github.com/NixOS/nixpkgs/tree/master/pkgs/misc/base16-builder
- https://github.com/NixOS/nixpkgs/tree/master/pkgs/tools/security/bitwarden-cli
- https://github.com/NixOS/nixpkgs/tree/master/pkgs/tools/networking/airfield
- https://github.com/NixOS/nixpkgs/tree/master/pkgs/development/web/remarkjs
- https://github.com/NixOS/nixpkgs/tree/master/pkgs/servers/rippled
- https://github.com/NixOS/nixpkgs/tree/master/pkgs/servers/web-apps/codimd

**Comment by doronbehar** [comment]
That was my original approach towards packaging this but @MetaDark has superseded me. I don't have a strong opinion as for the matter but you do point out an inconsistency..

**Comment by MetaDark** [comment]
@veprbl I was following the documentation in https://github.com/NixOS/nixpkgs/blob/master/doc/languages-frameworks/node.section.md

**Pull request merged**
